app/controllers/discourse_subscriptions/admin/subscriptions_controller.rb

```diff
@@ -11,42 +11,78 @@ class SubscriptionsController < ::Admin::AdminController
         begin
-          subscription_ids = Subscription.all.pluck(:external_id)
-          subscriptions = {
-            has_more: false,
-            data: [],
-            length: 0,
-            last_record: params[:last_record],
-          }
+          if is_stripe_configured?
+            discourse_customers = Customer.left_outer_joins(:user, :invite).all
 
-          if subscription_ids.present? && is_stripe_configured?
-            while subscriptions[:length] < PAGE_LIMIT
-              current_set = get_subscriptions(subscriptions[:last_record])
-
-              until valid_subscriptions =
-                      find_valid_subscriptions(current_set[:data], subscription_ids)
-                current_set = get_subscriptions(current_set[:data].last)
-                break if current_set[:has_more] == false
-              end
-
-              subscriptions[:data] = subscriptions[:data].concat(valid_subscriptions.to_a)
-              subscriptions[:last_record] = current_set[:data].last[:id] if current_set[
-                :data
-              ].present?
-              subscriptions[:length] = subscriptions[:data].length
-              subscriptions[:has_more] = current_set[:has_more]
-              break if subscriptions[:has_more] == false
+            subscriptions = get_subscription_data
+
+            subscriptions.each do |sub|
+              discourse_customer =
+                discourse_customers.filter { |c| c.customer_id == sub[:customer][:id] }.first
+
+              link_text(sub, discourse_customer)
             end
           elsif !is_stripe_configured?
             subscriptions = nil
           end
-
-          render_json_dump subscriptions
+          result = {
+            has_more: false,
+            data: subscriptions,
+            length: 0,
+            last_record: params[:last_record],
+          }
+          render_json_dump result
         rescue ::Stripe::InvalidRequestError => e
           render_json_error e.message
         end
       end
 
+      def update
+        user = ::User.find_by_username_or_email(params[:user])
+        if user.nil?
+          invite =
+            ::Invite.generate(
+              current_user,
+              email: params[:user],
+              skip_email: params[:skip_email],
+              custom_message: params[:custom_message],
+            )
+        end
+
+        subscription = ::Stripe::Subscription.retrieve(params[:id])
+
+        discourse_customer = Customer.find_by(customer_id: subscription[:customer])
+        if discourse_customer.nil?
+          discourse_customer = Customer.create(customer_id: subscription[:customer])
+        end
+
+        discourse_subscription = Subscription.find_by(external_id: params[:id])
+        if discourse_subscription.nil?
+          discourse_subscription =
+            Subscription.create(
+              customer_id: discourse_customer[:id],
+              external_id: subscription[:id],
+            )
+        end
```

Comment:
unless you don't care about the customer id when creating this record, you can do this.
```suggestion
Subscription.find_or_create_by(customer_id: customer[:id], external_id: subscription[:id])
```